File: src/data/sessions.ts
import { existsSync, readdirSync, statSync, readFileSync } from 'fs';
import { join, basename } from 'path';
import { homedir } from 'os';
import type { OcSession } from '../types.js';

const SESSIONS_DIR = join(homedir(), '.openclaw', 'agents', 'main', 'sessions');

const KNOWN_CHANNELS = ['whatsapp', 'telegram', 'slack', 'discord', 'web', 'voice', 'sms'];

function extractChannel(filePath: string): string {
  // Check filename: channel-sessionid.jsonl
  const stem = basename(filePath, '.jsonl');
  const dashIdx = stem.indexOf('-');
  if (dashIdx > 0) {
    const prefix = stem.slice(0, dashIdx).toLowerCase();
    if (KNOWN_CHANNELS.includes(prefix)) return prefix;
  }

  // Scan first few lines of the JSONL for a channel field
  try {
    const content = readFileSync(filePath, 'utf-8');
    for (const line of content.split('\n').slice(0, 8)) {
      if (!line.trim()) continue;
      try {
        const obj = JSON.parse(line);
        const ch = obj.channel ?? obj.metadata?.channel ?? obj.session?.channel;
        if (ch && typeof ch === 'string') return ch;
      } catch { continue; }
    }
  } catch { /* ignore */ }

  return 'main';
}

export function loadSessions(): OcSession[] {
  if (!existsSync(SESSIONS_DIR)) return [];

  const results: OcSession[] = [];

  try {
    const files = readdirSync(SESSIONS_DIR).filter(f => f.endsWith('.jsonl'));

    for (const file of files) {
      const filePath = join(SESSIONS_DIR, file);
      try {
        const stat = statSync(filePath);
        const sessionId = basename(file, '.jsonl');
        const sizeKb = Math.round(stat.size / 1024);
        const updatedAt = stat.mtimeMs;
        const date = new Date(updatedAt);
        const label = date.toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: '2-digit' });

        results.push({
          kind: 'session',
          id: `session#${sessionId}`,
          name: `${label}  ${sessionId.slice(0, 8)}`,
          channel: extractChannel(filePath),
          updatedAt,
          sessionFile: filePath,
          sizeKb,
        });
      } catch {
        // skip unreadable files
      }
    }
  } catch {
    return [];
  }

  // Most recent first
  return results.sort((a, b) => b.updatedAt - a.updatedAt);
}

File: internal/data/sessions.go
package data

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

var knownChannels = map[string]bool{
	"whatsapp": true, "telegram": true, "slack": true,
	"discord": true, "web": true, "voice": true, "sms": true,
}

func extractChannel(filePath string) string {
	stem := strings.TrimSuffix(filepath.Base(filePath), ".jsonl")
	if idx := strings.IndexByte(stem, '-'); idx > 0 {
		prefix := strings.ToLower(stem[:idx])
		if knownChannels[prefix] {
			return prefix
		}
	}

	f, err := os.Open(filePath)
	if err != nil {
		return "main"
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	for i := 0; i < 8 && scanner.Scan(); i++ {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var obj map[string]any
		if err := json.Unmarshal([]byte(line), &obj); err != nil {
			continue
		}
		if ch, ok := obj["channel"].(string); ok && ch != "" {
			return ch
		}
		if meta, ok := obj["metadata"].(map[string]any); ok {
			if ch, ok := meta["channel"].(string); ok && ch != "" {
				return ch
			}
		}
		if sess, ok := obj["session"].(map[string]any); ok {
			if ch, ok := sess["channel"].(string); ok && ch != "" {
				return ch
			}
		}
	}
	return "main"
}

func LoadSessions() []OcSession {
	home, _ := os.UserHomeDir()
	dir := filepath.Join(home, ".openclaw", "agents", "main", "sessions")

	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil
	}

	var items []OcSession
	for _, entry := range entries {
		if !strings.HasSuffix(entry.Name(), ".jsonl") {
			continue
		}
		filePath := filepath.Join(dir, entry.Name())
		info, err := entry.Info()
		if err != nil {
			continue
		}
		sessionID := strings.TrimSuffix(entry.Name(), ".jsonl")
		sizeKb := int(info.Size() / 1024)
		updatedAt := info.ModTime().UnixMilli()
		d := time.UnixMilli(updatedAt)
		label := fmt.Sprintf("%s %s  %s",
			d.Format("Jan"), fmt.Sprintf("%d", d.Day()),
			sessionID[:min(8, len(sessionID))],
		)

		items = append(items, OcSession{
			ID:          "session#" + sessionID,
			Name_:       label,
			Channel:     extractChannel(filePath),
			UpdatedAt:   updatedAt,
			SessionFile: filePath,
			SizeKb:      sizeKb,
		})
	}

	sort.Slice(items, func(i, j int) bool {
		return items[i].UpdatedAt > items[j].UpdatedAt
	})
	return items
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
